src/main/java/org/broadinstitute/hellbender/utils/haplotype/HaplotypeBAMWriter.java

```diff
@@ -0,0 +1,184 @@
+     */
+    public abstract void writeReadsAlignedToHaplotypes(final Collection<Haplotype> haplotypes,
+                                                       final GenomeLoc paddedReferenceLoc,
+                                                       final Collection<Haplotype> bestHaplotypes,
+                                                       final Set<Haplotype> calledHaplotypes,
+                                                       final ReadLikelihoods<Haplotype> readLikelihoods);
+
+    public void writeReadsAlignedToHaplotypes(final Collection<Haplotype> haplotypes,
+                                              final GenomeLoc paddedReferenceLoc,
+                                              final ReadLikelihoods<Haplotype> stratifiedReadMap) {
+        writeReadsAlignedToHaplotypes(haplotypes, paddedReferenceLoc, haplotypes, new HashSet<>(haplotypes), stratifiedReadMap);
+    }
+
+    /**
+     * Write out read aligned to haplotype to the BAM file
+     *
+     * @param originalRead the read we want to write aligned to the reference genome
+     */
+    protected void writeReadAgainstHaplotype(final GATKRead originalRead) {
+        output.add(originalRead);
+    }
+
+    /**
+     * Write out haplotypes as reads to the BAM, marking specifically those that are among the best haplotypes
+     *
+     * @param haplotypes a collection of haplotypes to write to the BAM
+     * @param bestHaplotypes a subset of haplotypes that contains those that are best "either good or called"
+     * @param paddedReferenceLoc the genome loc of the padded reference
+     */
+    protected void writeHaplotypesAsReads(final Collection<Haplotype> haplotypes,
+                                          final Set<Haplotype> bestHaplotypes,
+                                          final GenomeLoc paddedReferenceLoc) {
+        if (getWriteHaplotypes()) {
+            for (final Haplotype haplotype : haplotypes) {
+                writeHaplotype(haplotype, paddedReferenceLoc, bestHaplotypes.contains(haplotype));
+            }
+        }
+    }
+
+    /**
+     * Write out a representation of this haplotype as a read
+     *
+     * @param haplotype a haplotype to write out.  Cannot be null
+     * @param paddedRefLoc the reference location.  Cannot be null
+     * @param isAmongBestHaplotypes true if among the best haplotypes, false if it was just one possible but not so good
+     */
+    private void writeHaplotype(final Haplotype haplotype,
+                                final GenomeLoc paddedRefLoc,
+                                final boolean isAmongBestHaplotypes) {
+        final SAMRecord record = new SAMRecord(output.getOutHeader());
+        record.setReadBases(haplotype.getBases());
+        record.setAlignmentStart(paddedRefLoc.getStart() + haplotype.getAlignmentStartHapwrtRef());
+        record.setBaseQualities(Utils.dupBytes((byte) '!', haplotype.getBases().length));
```

Comment:
add a comment - `!` is value 33. The character is picked because of what it looks like and not the value.